firedrake/mg/embedded.py

```diff
@@ -265,37 +274,44 @@ def restrict(self, gf, gc):
         """
         Vc = gc.function_space()
         Vf = gf.function_space()
-
         source_element = Vf.ufl_element()
         target_element = Vc.ufl_element()
+
+        cache = self.cache(source_element)
```

Comment:
~This block could definitely do with a comment.~ A sensible function would avoid this requirement.

Reply:
And could the whole thing be refactored into a function:
```py
def requires_update(self, element, op, uc, uf):
    key = (op, uc, uf)
    ...
    return dat_versions != cache._dat_versions[key]
```

The method name could probably be made even clearer than this. Also include a docstring.